### ISSUE TITLE
Beware: swat TK may conflict with sas foundation TK!

### ISSUE THREAD

**Issue by gygabyte017** [sassoftware/python-swat#63]
Hi, I want to report an issue that drove me crazy for days.

On my windows machine, I have a standard SAS 9.4 foundation, and I also have installed latest swat library. In my code, I need both to operate on CAS (via the swat package) and to launch some sas base program via command line ("sas.exe -sysin script.sas"). Well, after the swat package has loaded, the sas.exe program fails to start "Unable to load extension tksecure.dll".

I found out that sas.exe needs TK libraries to start, and if the swat package is loaded the default sas.exe path is overwritten and this caused in my case a version mismatch error (hence the load extension failure).

The solution is to manually tell swat to ignore its own TK library and instead use the sas foundation one:

`swat.options.tkpath = r'C:\SAS\SASFoundation\9.4\core\sasext'`

Perhaps it's worth it to add a note to the documentation in case others have both SAS 9.4 foundation and swat installed on the same Windows machine.

**Comment by kesmit13** [comment]
I'd like to get some more information about this issue because I'm not sure why there would be a collision of the two programs.  SWAT installs its own version of TK in the Python library tree which shouldn't have an effect on any other program.

My first question is, are you running both sas and swat in the same window or different windows?

**Comment by gygabyte017** [comment]
Sure, let me add some details.

First, the python code I'm running in the same python kernel:

```
import swat, os
conn = swat.CAS(host, port, username, password)
os.system("r'C:\SAS\SASFoundation\9.4\sas.exe -sysin program.sas")
```

As I said before, on the same machine I have a SAS 9.4M6 full installation, and this installation brings the TK libraries in C:\SAS\SASFoundation\9.4\core\sasext.
Now, with swat installed, there is another set of TK libraries in its own package folder of the python distribution. I checked out just for curiosity, and it seems that the TK libraries installed by SAS Foundation are a lot more of the one installed by swat.

Now, if I run that code, sas.exe refuses to start saying something like: "Critical Error loading host supervisor; TK could not initialize at tkboot; failed to load tksecure extension". If I remove the swat import and restart the kernel, sas.exe works as expected.

So, after a a lot of attempts, I ran ProcMon trying to see if I spotted something. And I did. I saw that without the swat import, sas.exe loads some TK libraries from its own path C:\SAS\SASFoundation\9.4\core\sasext and starts correctly. With the swat import, sas.exe begins to load the same TK libraries this time from the swat own package folder, and then it comes to tksecure.dll: this dll is not present in the swat TK libraries package folder, so sas.exe begins to scan now its own folder C:\SAS\SASFoundation\9.4\core\sasext, it founds it, and it dies with that message. I saw also that it writes a log in a temp folder, so I traced the log and inside it there were just one line saying something like: "Cannot load tksecure.dll, version mismatch".

Hence, I figured out that the problem is that the two TK libraries have different versions. sas.exe alone loads its version, and that's ok; swat alone loads its version, and that's ok; but when I use swat AND sas.exe in the same code, sas.exe loads some dll from swat folder and some other from its own folder, and since the versions are different, it crashes!

This is all I understood :-)

**Comment by kesmit13** [comment]
Thanks.  That information is very helpful.  It may be that the TK subsystem in SWAT is setting environment variables that are being picked up by sas.exe when it starts (I'll have to look into that some more).  It's true, the two TK systems do contain a different set of libraries.  SWAT only needs a small subset, whereas sas.exe needs everything available.  We've tried to optimized the SWAT package size by removing things that weren't necessary.

I wonder if using something like `subprocess.run(['sas.exe', '-sysin', 'program.sas'], env={}, shell=False)` would work since it cleans the environment before running the command.  It may be that it introduces other issues though.

**Comment by kesmit13** [comment]
BTW, I wanted to comment on mixing the TK versions.  It may be that you are getting lucky and the libraries just happen to be compatible, but it's quite possible that mixing library versions will result in a SWAT crash somewhere.  I realize that you were just looking for a solution and found one that worked, but I wanted to mention it in case you saw any other odd behavior.

**Comment by gygabyte017** [comment]
Yes, I believe as well that it's just a matter of environment variables about the path where to look for TK dll.

And yes, you're right I have thought that since swat is using a different set ok TK libraries now it may crash anytime, however so far things like connection, upload_frame, basic querying, merging, dropping, promoting, ecc seem to work well. Maybe swat works because the version mismatch is just a minor difference, since in our company we usually have the latest Viya and Foundation asap, and I believe swat TKs also are like the latest version. But when it comes to sas.exe dll loading, I believe the dll version signature checking is very strict and if it is not exactly the same as expected it refuses to load.

It would be great however to have control about what uses what: let me know if I can help somehow with some testing, since I have this "weird" setup on my machine!

**Comment by kesmit13** [comment]
Here's something you can try.  After making your CAS connection, do the following:
```
del os.environ['TKPATH']
```
On Windows, SWAT sets that variable, but it's not needed after TK has started.

**Comment by kesmit13** [comment]
After trying the above myself, it looks like Python isn't seeing the environment variable to delete it.  Maybe setting it to an alternate value would work.
```
os.environ['TKPATH'] = os.pathsep
```
If you can give this a try, I'd be interested in know if it works.  This fix could be easily added to the Python source.

**Comment by gygabyte017** [comment]
I apologize, I realized that I said something wrong in my previous message.

This code:
```
import os, swat
conn = swat.CAS(...)
os.system("sas.exe")
```

works correctly, either the CAS connection and the sas.exe execution.

Instead, this is the code that doesn't work:
```
import os, swat
os.system("sas.exe")
```

But this works (obiouvsly):
```
import os
os.system("sas.exe")
```

So it seems that just importing the swat package causes the env paths to be changed. However, after a connection has been established, the env paths seems to be reset to the previous values.

**Comment by kesmit13** [comment]
Hmm, that's even more interesting.  Could you try the following and see if this works?
```
import os, swat
os.environ['TKPATH'] = ''
os.system('sas.exe')
```
Or possibly:
```
import os, swat
os.environ['TKPATH'] = os.pathsep
os.system('sas.exe')
```
If I can find a solution on the Python side, I should be able to patch it fairly quickly.  Changes in the C side can take more time to show up in a release.

**Comment by gygabyte017** [comment]
First one: same bad result.
Second one: it works!

**Comment by kesmit13** [comment]
Ok, one more check and I think we might have something.  As long as swat still works after setting the TKPATH manually, I think we have a fix.
```
import os, swat
os.environ['TKPATH'] = os.pathsep
os.system('sas.exe')
conn = swat.CAS(...)
```

**Comment by gygabyte017** [comment]
I confirm this way both sas.exe and cas connection work (and vice versa, I tried also swapping the 3rd and 4th line, you never know :-)). So yes I believe we have a fix!

**Comment by kesmit13** [comment]
Fixed in c0053d4f19bb29032f1d73bfdff010190e02a015
